File: 2064/solution.java
class Solution {
    public int minimizedMaximum(int n, int[] quantities) {
        int l = 1;
        int r = 0;
        for (int q : quantities) {
            r = Math.max(r, q);
        }
        int res = r;
        while (l <= r) {
            int p = l + (r - l) / 2;
            if (canSplit(p, n, quantities)) {
                res = p;
                r = res - 1;
            } else {
                l = p + 1;
            }
        }
        return res;

    }

    private boolean canSplit(int maxCapacity, int n, int[] quantities) {
        int res = 0;
        for (int q : quantities) {
            res += q / maxCapacity + (q % maxCapacity == 0 ? 0 : 1);
        }
        return res <= n;
    }
}

// time O(log(max(n))*n)
// space O(1)

File: 2064/solution.py
from math import ceil


class Solution:
    def minimizedMaximum(self, n: int, quantities: list[int]) -> int:
        def can_split(max_val):
            return sum(ceil(num / max_val) for num in quantities) <= n

        l, r = 1, max(quantities)
        res = None
        while l <= r:
            p = (l + r) // 2
            if can_split(p):
                res = p
                r = p - 1
            else:
                l = p + 1
        return res
